### ISSUE TITLE
Failed to load plugin /usr/lib/midimonster/lua.so: ... undefined symbol: lua_gettop

### ISSUE THREAD

**Issue by mxmilkb** [cbdevnet/midimonster#36]
Currently I get a failure when trying to run midimonster;

```
% midimonster /usr/share/midimonster/evdev.cfg
Failed to load plugin /usr/lib/midimonster/lua.so: /usr/lib/midimonster/lua.so: undefined symbol: lua_gettop
Cannot configure unknown backend artnet
Failed to read configuration file /usr/share/midimonster/evdev.cfg
MIDIMonster v0.1
Usage:
	midimonster <configfile>
% ll /usr/lib/midimonster/lua.so
-rwxr-xr-x 1 root root 18K Dec  3 16:08 /usr/lib/midimonster/lua.so
```

Arch Linux, being rolling release, currently has lua 5.3.5-2, though I also have lua51 and lua52 installed.

From the build output if it helps;
```
...
make[1]: Entering directory '/home/milk/state/pkgs/in_progress/midimonster-git/src/midimonster/backends'
rm -f libmmbackend.o artnet.so osc.so loopback.so sacn.so lua.so maweb.so jack.so midi.so evdev.so ola.so artnet.dll osc.dll loopback.dll sacn.dll maweb.dll winmidi.dll
make[1]: Leaving directory '/home/milk/state/pkgs/in_progress/midimonster-git/src/midimonster/backends'
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -DPLUGINS=\"/usr/lib/midimonster\" -D_FORTIFY_SOURCE=2  -c -o config.o config.c
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -DPLUGINS=\"/usr/lib/midimonster\" -D_FORTIFY_SOURCE=2  -c -o backend.o backend.c
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -DPLUGINS=\"/usr/lib/midimonster\" -D_FORTIFY_SOURCE=2  -c -o plugin.o plugin.c
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -DPLUGINS=\"/usr/lib/midimonster\" -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,-export-dynamic midimonster.c config.o backend.o plugin.o -ldl -o midimonster
make -C backends
make[1]: Entering directory '/home/milk/state/pkgs/in_progress/midimonster-git/src/midimonster/backends'
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic -D_FORTIFY_SOURCE=2 -g -fPIC -I../  -c -o libmmbackend.o libmmbackend.c
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic  artnet.c libmmbackend.o -o artnet.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic  osc.c libmmbackend.o -o osc.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic  loopback.c  -o loopback.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic  sacn.c libmmbackend.o -o sacn.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
Package lua5.3 was not found in the pkg-config search path.
Perhaps you should add the directory containing `lua5.3.pc'
to the PKG_CONFIG_PATH environment variable
Package 'lua5.3', required by 'virtual:world', not found
Package lua5.3 was not found in the pkg-config search path.
Perhaps you should add the directory containing `lua5.3.pc'
to the PKG_CONFIG_PATH environment variable
Package 'lua5.3', required by 'virtual:world', not found
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic   lua.c  -o lua.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
Package lua5.3 was not found in the pkg-config search path.
Perhaps you should add the directory containing `lua5.3.pc'
to the PKG_CONFIG_PATH environment variable
Package 'lua5.3', required by 'virtual:world', not found
lua.c:55:17: warning: ‘lua_interval’ defined but not used [-Wunused-function]
   55 | static uint32_t lua_interval(){
      |                 ^~~~~~~~~~~~
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic -lssl maweb.c libmmbackend.o -o maweb.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
maweb.c: In function ‘maweb_handle’:
maweb.c:803:3: warning: ‘bytes_handled’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  803 |   memmove(data->buffer, data->buffer + bytes_handled, (data->offset + bytes_read) - bytes_handled);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
maweb.c:754:67: note: ‘bytes_handled’ was declared here
  754 |  ssize_t bytes_read, bytes_left = data->allocated - data->offset, bytes_handled;
      |                                                                   ^~~~~~~~~~~~~
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic -ljack -lpthread jack.c  -o jack.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic -lasound midi.c  -o midi.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden -g -fPIC -I../ -Wall -Wpedantic -I/usr/include/libevdev-1.0/  -levdev  evdev.c  -o evdev.so -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -shared
evdev.c: In function ‘evdev_find’:
evdev.c:158:4: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
  158 |    strncpy(result, file_path, strlen(file_path));
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: Leaving directory '/home/milk/state/pkgs/in_progress/midimonster-git/src/midimonster/backends'
...
```

**Comment by keldu** [comment]
Either debian or archlinux maintainer mismanaged the packaging of the lua pkgconf files.  
I think the Archlinux maintainer should change their naming scheme since all other packages I currently have overflown have points in their version string.  

On debian it's probably in '/usr/lib/pkgconfig/lua5.3.pc' (Don't have a debian machine with lua installed).  
On arch it's '/usr/lib/pkgconfig/lua53.pc'  

So you either create a softlink or modify the makefile in backends to resolve your problem in the short run.  
'grep -rni "pkg-config --libs lua5.3"  '

This situation would be resolved by contacting the package maintainer on Arch and correcting that mistake.

**Comment by keldu** [comment]
Apparently Debian has `lua53.pc` AND `lua5.3.pc`

Still strange that `lua` diverges from the other `pc` files

**Comment by keldu** [comment]
The build should fail if it doesn't find the package.......  
For now I think it's better to use `lua53.pc` instead of `lua5.3` in the makefile and the build should fail if it doesn't find a package with pkgconf.

**Comment by cbdevnet** [comment]
Researching this brings up a lengthy [thread on the lua-users mailing list](http://lua-users.org/lists/lua-l/2015-03/msg00293.html) about this.

Personally, I think the pkg-config file should be named 'lua5.3' (as it was assumed in the Makefile up until now) because that makes clear the exact version requested (if Lua were at some point to release a Version 53, 'lua53' would be ambiguous). In the interest of interoperability (which is kind of the point of the whole pkg-config thing, much as I dislike it), I changed it to 'lua53' in a0831a2b970404eaa9b80ff97ab46ee759131414. I also added checks to the invocations in the makefile which stop the build if an error is encountered.

Looking at the CI report, the change breaks the build on OSX. Great.
